### PR TITLE
Add NumericUpDown control

### PR DESCRIPTION
**Pull request by ConradRoesch**
Note that the Value property is of type Nullable<decimal>, since the underlying Avalonia control was recently reworked to treat an empty entry as null (https://github.com/AvaloniaUI/Avalonia/pull/8259). 

In a Fabulous context, is a nullable Value property the desired behavior for this control? Unlike the Date and Time picker controls, a non-nullable value for NumericUpDown cannot be enforced solely by the constructor, since the user has the option to manually clear the input via keyboard entry. (Note also that watermark visibility is contingent on the Value property being nullable.)

### CODE REVIEW

File: src/Fabulous.Avalonia/Widgets/Controls/NumericUpDown.fs [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+    let TextConverter =
+        Attributes.defineAvaloniaPropertyWithEquality NumericUpDown.TextConverterProperty
+
+    let Watermark =
+        Attributes.defineAvaloniaPropertyWithEquality NumericUpDown.WatermarkProperty
+
+    let Value =
+        Attributes.defineAvaloniaPropertyWithEquality NumericUpDown.ValueProperty
+
+    let ValueChanged =
```
Comment (edgarfgp) [inline src/Fabulous.Avalonia/Widgets/Controls/NumericUpDown.fs:65]: See my previous comment. For this one, we can use `Attributes.defineAvaloniaPropertyWithChangedEvent` with converters to massage the value :) 
```fsharp
let ValueChanged =
        Attributes.defineAvaloniaPropertyWithChangedEvent
            "NumericUpDown_ValueChanged"
            NumericUpDown.ValueProperty
            Option.toNullable
            Option.ofNullable
```

File: src/Fabulous.Avalonia/Widgets/Controls/NumericUpDown.fs [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+        Attributes.defineAvaloniaPropertyWithChangedEvent' "NumericUpDown_ValueChanged" NumericUpDown.ValueProperty
+
+    let LostFocus =
```
Comment (edgarfgp) [inline src/Fabulous.Avalonia/Widgets/Controls/NumericUpDown.fs:68]: I believe this one should not be here. This is part of the Interactive. fs Pending TODO. 

As NumericUpDown inherits <- TemplatedControl <- Control <- InputElement <- `Interactive`. We should be able to access to the events there

File: src/Fabulous.Avalonia/Widgets/Controls/NumericUpDown.fs [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+        Attributes.defineEvent<RoutedEventArgs> "NumericUpDown_LostFocus" (fun target ->
+            (target :?> NumericUpDown).LostFocus)
+
+[<AutoOpen>]
+module NumericUpDownBuilders =
+    type Fabulous.Avalonia.View with
+
+        static member inline NumericUpDown<'msg>(value: Nullable<decimal>, valueChanged: Nullable<decimal> -> 'msg) =
```
Comment (edgarfgp) [inline src/Fabulous.Avalonia/Widgets/Controls/NumericUpDown.fs:76]: See my previous comments. This now can we a `decimal option` and be more F# friendly :) 

```fsharp
static member inline NumericUpDown<'msg>(value: decimal option, valueChanged: decimal option -> 'msg) =
            WidgetBuilder<'msg, IFabNumericUpDown>(
                NumericUpDown.WidgetKey,
                NumericUpDown.Value.WithValue(value),
                NumericUpDown.ValueChanged.WithValue(ValueEventData.create value (fun args -> valueChanged args |> box))
            )
```

File: src/Fabulous.Avalonia/Widgets/Controls/NumericUpDown.fs [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+
+    [<Extension>]
+    static member inline minimum(this: WidgetBuilder<'msg, #IFabNumericUpDown>, value: decimal) =
+        this.AddScalar(NumericUpDown.Minimum.WithValue(value))
+
+    [<Extension>]
+    static member inline numberFormat
+        (
+            this: WidgetBuilder<'msg, #IFabNumericUpDown>,
+            value: System.Globalization.NumberFormatInfo
```
Comment (edgarfgp) [inline src/Fabulous.Avalonia/Widgets/Controls/NumericUpDown.fs:138]: We could `open System.Globalization` at the top if possible

File: src/Fabulous.Avalonia/Widgets/Controls/NumericUpDown.fs [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+        ) =
+        this.AddScalar(NumericUpDown.NumberFormat.WithValue(value))
+
+    [<Extension>]
+    static member inline parsingNumberStyle
+        (
+            this: WidgetBuilder<'msg, #IFabNumericUpDown>,
+            value: System.Globalization.NumberStyles
```
Comment (edgarfgp) [inline src/Fabulous.Avalonia/Widgets/Controls/NumericUpDown.fs:146]: We could `open System.Globalization` at the top if possible

File: src/Fabulous.Avalonia/Widgets/Controls/NumericUpDown.fs [hunk 6]
```diff
@@ -0,0 +1,176 @@
...
+
+    [<Extension>]
+    static member inline text(this: WidgetBuilder<'msg, #IFabNumericUpDown>, value: string) =
+        this.AddScalar(NumericUpDown.Text.WithValue(value))
+
+    [<Extension>]
+    static member inline textConverter
+        (
+            this: WidgetBuilder<'msg, #IFabNumericUpDown>,
+            value: Avalonia.Data.Converters.IValueConverter
```
Comment (edgarfgp) [inline src/Fabulous.Avalonia/Widgets/Controls/NumericUpDown.fs:162]: We could `open Avalonia.Data.Converters` at the top if possible

**Review by edgarfgp — CHANGES_REQUESTED** [review]
Thanks for this 👍.  Just couple comments

File: src/Fabulous.Avalonia/Widgets/Controls/NumericUpDown.fs [hunk 7]
```diff
@@ -0,0 +1,176 @@
...
+
+    let Text = Attributes.defineAvaloniaPropertyWithEquality NumericUpDown.TextProperty
+
+    let TextConverter =
+        Attributes.defineAvaloniaPropertyWithEquality NumericUpDown.TextConverterProperty
+
+    let Watermark =
+        Attributes.defineAvaloniaPropertyWithEquality NumericUpDown.WatermarkProperty
+
+    let Value =
```
[the remark this answers is not in the document]
Comment (ConradRoesch) [reply src/Fabulous.Avalonia/Widgets/Controls/NumericUpDown.fs:62]: Thank you for providing the converter code along with the explanation (my background is C#, so I'm still learning many of the features baked into F#). I'm sure I'll see this converter again as we work through the remaining controls!

[the remark this answers is not in the document]
Comment (edgarfgp) [reply src/Fabulous.Avalonia/Widgets/Controls/NumericUpDown.fs:62]: No worries,  happy to help :). Tim did the same for me a year ago when I started with F# and Fabulous.

**Review by edgarfgp — APPROVED** [review]

**Comment by edgarfgp** [comment]
Awesome 👍 . Thanks

**Pull request merged**